__tests__/Auth0Client.test.ts

```diff
@@ -173,28 +173,389 @@ describe('Auth0Client', () => {
+      assertPost('https://auth0_domain/oauth/token', {
+        redirect_uri: TEST_REDIRECT_URI,
+        client_id: TEST_CLIENT_ID,
+        code_verifier: TEST_CODE_VERIFIER,
+        grant_type: 'authorization_code',
+        code: 'my_code'
+      });
+    });
+
+    it('opens popup with custom auth0Client', async () => {
+      const auth0Client = { name: '__test_client_name__', version: '9.9.9' };
+      const auth0 = await setup({ auth0Client });
+
+      await loginWithPopup(auth0);
+
+      expect(mockWindow.open).toHaveBeenCalled();
+      const [[url]] = (<jest.Mock>mockWindow.open).mock.calls;
+      assertUrlEquals(url, TEST_DOMAIN, '/authorize', {
+        auth0Client: btoa(JSON.stringify(auth0Client))
+      });
+    });
+
+    it('throws error if state from popup response is different from the provided state', async () => {
+      authorizationResponse.state = 'other-state';
+
+      const auth0 = setup();
+
+      await expect(loginWithPopup(auth0)).rejects.toThrowError('Invalid state');
+    });
+
+    it('calls `tokenVerifier.verify` with the `issuer` from in the oauth/token response', async () => {
+      const auth0 = setup({
+        issuer: 'test-123.auth0.com'
+      });
+
+      await loginWithPopup(auth0);
+      expect(tokenVerifier).toHaveBeenCalledWith(
+        expect.objectContaining({
+          iss: 'https://test-123.auth0.com/'
+        })
+      );
+    });
+    it('calls `tokenVerifier.verify` with the `leeway` from constructor', async () => {
+      const auth0 = setup({ leeway: 10 });
+
+      await loginWithPopup(auth0);
+
+      expect(tokenVerifier).toHaveBeenCalledWith(
+        expect.objectContaining({
+          leeway: 10
+        })
+      );
+    });
+    it('calls `tokenVerifier.verify` with undefined `max_age` when value set in constructor is an empty string', async () => {
+      const auth0 = setup({ max_age: '' });
+
+      await loginWithPopup(auth0);
+
+      expect(tokenVerifier).toHaveBeenCalledWith(
+        expect.objectContaining({
+          max_age: undefined
+        })
+      );
+    });
+    it('calls `tokenVerifier.verify` with the parsed `max_age` string from constructor', async () => {
+      const auth0 = setup({ max_age: '10' });
+
+      await loginWithPopup(auth0);
+
+      expect(tokenVerifier).toHaveBeenCalledWith(
+        expect.objectContaining({
+          max_age: 10
+        })
+      );
+    });
+    it('calls `tokenVerifier.verify` with the parsed `max_age` number from constructor', async () => {
+      const auth0 = setup({ max_age: 10 });
+
+      await loginWithPopup(auth0);
+
+      expect(tokenVerifier).toHaveBeenCalledWith(
+        expect.objectContaining({
+          max_age: 10
+        })
+      );
+    });
+
+    it('saves into cache', async () => {
+      const auth0 = setup();
+
+      jest.spyOn(auth0['cache'], 'save');
+
+      await loginWithPopup(auth0);
+
+      expect(auth0['cache']['save']).toHaveBeenCalledWith(
```

Comment:
Because I wanted to limit mocking to only what was necessary, I didn't mock the caching.
However, as `cache` is a private property on `Auth0Client`, we need to access it as above in order to spyOn it or use it to expect it has been called.

Reply:
I wonder if there's a way we can verify the cache behaviour without relying on internal implementation 🤔  I agree that it shouldn't be mocked. If we were using local storage it would be easier as we already have helpers for that.

However, the spy syntax here is nice and simple 🤷  what about a helper that calls something like `getTokenSilently` and verifies that no call was made to `/authorize`? That would show that a value was cached.

Another question is, is this a useful test?

Reply:
Calling getTokenSilently wouldnt test that the access_token was cached, but I understand the reasoning about the rest. I like that idea as that tests less implementation  but focusses on why we are caching it.

About the usefulness of this test, good question. Maybe it tests to much of the implementation ad we are better off adding a test for getTokenSilently the way u describe.

Reply:
As discussed, keeping this as is due to the difficulty to spy on the cache without mocking it.